plugin/pkg/scheduler/generic_scheduler.go

```diff
@@ -170,40 +170,72 @@ func PrioritizeNodes(pod *api.Pod, podLister algorithm.PodLister, priorityConfig
 		return EqualPriority(pod, podLister, nodeLister)
 	}
 
+	var (
+		mu  sync.Mutex
+		err error
+	)
 	combinedScores := map[string]int{}
+	errC := make(chan error)
+
 	for _, priorityConfig := range priorityConfigs {
-		weight := priorityConfig.Weight
 		// skip the priority function if the weight is specified as 0
-		if weight == 0 {
+		if priorityConfig.Weight == 0 {
 			continue
 		}
-		priorityFunc := priorityConfig.Function
-		prioritizedList, err := priorityFunc(pod, podLister, nodeLister)
-		if err != nil {
-			return schedulerapi.HostPriorityList{}, err
-		}
-		for _, hostEntry := range prioritizedList {
-			combinedScores[hostEntry.Host] += hostEntry.Score * weight
+		go func(algorithm.PriorityConfig) {
+			weight := priorityConfig.Weight
+			priorityFunc := priorityConfig.Function
+			prioritizedList, err := priorityFunc(pod, podLister, nodeLister)
+			if err != nil {
+				errC <- err
+				return
+			}
+			for _, hostEntry := range prioritizedList {
+				mu.Lock()
+				combinedScores[hostEntry.Host] += hostEntry.Score * weight
+				mu.Unlock()
```

Comment:
I think locking & unlocking in every single look is very inefficient.
Also, making a copy of every element in the list doesn't make much sense to.

I suggest changing to:

mu.Lock()
for i := range prioritizedList {
  combinedScores[prioritizedList[i].Host] += hostEntry.Score * weight
}
mu.Unlock()